### PR TITLE
HOWTO: Implementing a Squeeze and Excite block

### LINKED ISSUE #873: Add HOWTO: Implementing a Squeeze and Excite block
pre-Linen code (should be easy to port to Linen):

```
class SEBlock(nn.Module):
  """Applies Squeeze-and-Excitation."""

  def apply(self, x, hidden_size, act_fn=nn.relu, axis=(1, 2), dtype=jnp.float32):
    y = x.mean(axis=axis, keepdims=True)
    y = nn.Dense(
        y, features=hidden_size,
        dtype=dtype, name='reduce')
    y = act_fn(y)
    y = nn.Dense(
        y, features=x.shape[-1],
        dtype=dtype, name='expand')
    return nn.sigmoid(y) * x
```

### PR DESCRIPTION
**Pull request by joaogui1**
Closes #873

### CODE REVIEW

**Comment by joaogui1** [comment]
Hey @marcvanzee can you review this please?

File: docs/howtos/squeeze_and_excite.rst [hunk 1]
```diff
@@ -0,0 +1,32 @@
+Squeeze and Excite Block
+=============================
+
+We will show you how to...
+
+* Implement a Squeeze and Excite Block
+Squeeze and Excite Blocks were introduced in `"Squeeze-and-Excitation Networks" <https://arxiv.org/abs/1709.01507>`_ seek to
```
Comment (8bitmp3) [inline docs/howtos/squeeze_and_excite.rst:7]: In the paper: "Squeeze-and- Excitation (SE) block"

So, how about this:

```suggestion
Squeeze-and- Excitation blocks `(Hu at el., 2017) <https://arxiv.org/abs/1709.01507>`_ seek to
```

File: docs/howtos/squeeze_and_excite.rst [hunk 2]
```diff
@@ -0,0 +1,32 @@
+Squeeze and Excite Block
```
Comment (8bitmp3) [inline docs/howtos/squeeze_and_excite.rst:1]: In the paper: "Squeeze-and- Excitation (SE) block"

Suggestion:

```suggestion
Squeeze-and- Excitation (SE) Block
```

File: docs/howtos/squeeze_and_excite.rst [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+=============================
+
+We will show you how to...
+
+* Implement a Squeeze and Excite Block
+Squeeze and Excite Blocks were introduced in `"Squeeze-and-Excitation Networks" <https://arxiv.org/abs/1709.01507>`_ seek to
+improve the quality of representations learned by a network by modelling the dependencies between channels of convolutional features.
+The first step in the block is the Squeeze operation: all channels are aggregated by taking their means, changing the 
+tensor's shape from (H, W, C) to (1, 1, C). Then it goes through an Excitation operation: the mean vector passes through a 2 layer MLP
```
Comment (8bitmp3) [inline docs/howtos/squeeze_and_excite.rst:10]: Suggestion:

- I recommend clarifying what "it' means here ("Then, it goes through....") since there are different moving parts

File: docs/howtos/squeeze_and_excite.rst [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+and then is used in a self-gating mechanism, where it modulates each channel. The output still has shape (H, W, C) and so can just be passed
```
Comment (8bitmp3) [inline docs/howtos/squeeze_and_excite.rst:11]: Suggestion:

To avoid the "then... then" repetition, how about "then... before being..."

```suggestion
before being  used in a self-gating mechanism, where it modulates each channel. The output still has shape (H, W, C) and so can just be passed
```

Comment (8bitmp3) [inline docs/howtos/squeeze_and_excite.rst:11]: Suggestion:

- Minor refactoring

```suggestion
and then is used in a self-gating mechanism, where it modulates each channel. The output retains the shape (H, W, C) and therefore can just be passed
```

File: docs/howtos/squeeze_and_excite.rst [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+on to the next layer.
+.. code-block:: python
+
+
+  class SEBlock(nn.Module):
+    """Squeeze-and-Excitation Block."""
```
Comment (8bitmp3) [inline docs/howtos/squeeze_and_excite.rst:17]: Suggestion:

```suggestion
    """The Squeeze-and-Excitation block."""
```

**Review by 8bitmp3 — COMMENTED** [review]
This is awesome @joaogui1. I've provided a few minor suggestions in case you're interested.

File: docs/howtos/squeeze_and_excite.rst [hunk 6]
```diff
@@ -0,0 +1,32 @@
+Squeeze and Excite Block
+=============================
+
+We will show you how to...
+
+* Implement a Squeeze and Excite Block
```
Comment (8bitmp3) [inline docs/howtos/squeeze_and_excite.rst:6]: In the paper: "Squeeze-and- Excitation (SE) block"

```suggestion
* Implement a Squeeze-and- Excitation (SE) block

```

File: docs/howtos/squeeze_and_excite.rst (hunk 3 above)
Comment (joaogui1) [reply docs/howtos/squeeze_and_excite.rst]: Changed it

File: docs/howtos/squeeze_and_excite.rst [hunk 7]
```diff
@@ -0,0 +1,33 @@
+Squeeze-and- Excitation (SE) Block
```
Comment (8bitmp3) [inline docs/howtos/squeeze_and_excite.rst:1]: Whoops! Bad copy-paste on my part. Here's the correction

```suggestion
Squeeze-and-Excitation Block
```

File: docs/howtos/squeeze_and_excite.rst [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+=============================
+
+We will show you how to...
+
+* Implement a Squeeze-and- Excitation (SE) block
```
Comment (8bitmp3) [inline docs/howtos/squeeze_and_excite.rst:6]: Whoops! Bad copy-paste on my part. Here's the correction

```suggestion
* Implement a Squeeze-and-Excitation block
```

File: docs/howtos/squeeze_and_excite.rst [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+Squeeze-and- Excitation blocks `(Hu at el., 2017) <https://arxiv.org/abs/1709.01507>`_ seek to
```
Comment (8bitmp3) [inline docs/howtos/squeeze_and_excite.rst:8]: Whoops! Bad copy-paste on my part. Here's the correction

```suggestion
Squeeze-and-Excitation blocks `(Hu at el., 2017) <https://arxiv.org/abs/1709.01507>`_ seek to
```

**Comment by joaogui1** [comment]
Just pining this @8bitmp3

**Comment by joaogui1** [comment]
@8bitmp3 @marcvanzee can you check this out?

File: docs/howtos/squeeze_and_excite.rst [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+Squeeze-and-Excitation blocks `(Hu at el., 2017) <https://arxiv.org/abs/1709.01507>`_ seek to
+improve the quality of representations learned by a network by modelling the dependencies between channels of convolutional features.
+The first step in the block is the Squeeze operation: a copy of all channels are aggregated by taking their means, generating 
+a tensor with shape (1, 1, C). Then we have the Excitation operation: the mean vector passes through a 2 layer MLP
+before being  used in a self-gating mechanism, where it modulates each channel of the original block. 
+The output retains the shape (H, W, C) and therefore can just be passed on to the next layer.
+.. code-block:: python
+
+
```
Comment (marcvanzee) [inline docs/howtos/squeeze_and_excite.rst:16]:
```suggestion

.. testcode::

```

File: docs/howtos/squeeze_and_excite.rst [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+=============================
+
+We will show you how to...
+
+* Implement a Squeeze-and-Excitation block
+
+Squeeze-and-Excitation blocks `(Hu at el., 2017) <https://arxiv.org/abs/1709.01507>`_ seek to
+improve the quality of representations learned by a network by modelling the dependencies between channels of convolutional features.
+The first step in the block is the Squeeze operation: a copy of all channels are aggregated by taking their means, generating 
+a tensor with shape (1, 1, C). Then we have the Excitation operation: the mean vector passes through a 2 layer MLP
```
Comment (marcvanzee) [inline docs/howtos/squeeze_and_excite.rst:11]:
```suggestion
a tensor with shape ``(1, 1, C)``. Then we have the Excitation operation: the mean vector passes through a 2 layer MLP
```

File: docs/howtos/squeeze_and_excite.rst [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+before being  used in a self-gating mechanism, where it modulates each channel of the original block.
```
Comment (marcvanzee) [inline docs/howtos/squeeze_and_excite.rst:12]:
```suggestion
before being used in a self-gating mechanism, where it modulates each channel of the original block. 
```

File: docs/howtos/squeeze_and_excite.rst [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+We will show you how to...
+
+* Implement a Squeeze-and-Excitation block
+
+Squeeze-and-Excitation blocks `(Hu at el., 2017) <https://arxiv.org/abs/1709.01507>`_ seek to
+improve the quality of representations learned by a network by modelling the dependencies between channels of convolutional features.
+The first step in the block is the Squeeze operation: a copy of all channels are aggregated by taking their means, generating 
+a tensor with shape (1, 1, C). Then we have the Excitation operation: the mean vector passes through a 2 layer MLP
+before being  used in a self-gating mechanism, where it modulates each channel of the original block. 
+The output retains the shape (H, W, C) and therefore can just be passed on to the next layer.
```
Comment (marcvanzee) [inline docs/howtos/squeeze_and_excite.rst:13]:
```suggestion
The output retains the shape ``(H, W, C)`` and therefore can just be passed on to the next layer.
```

File: docs/howtos/squeeze_and_excite.rst [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    def __call__(self, x):
+        y = x.mean(axis=axis, keepdims=True)
+        y = nn.Dense(
+            y, features=self.hidden_size,
+            name='reduce')
+        y = act(y)
+        y = nn.Dense(
+            y, features=x.shape[-1],
+            name='expand')
+        return nn.sigmoid(y) * x
```
Comment (marcvanzee) [inline docs/howtos/squeeze_and_excite.rst:33]:
```suggestion
      y = x.mean(axis=axis, keepdims=True)
      y = nn.Dense(
          y, features=self.hidden_size,
          name='reduce')
      y = act(y)
      y = nn.Dense(
          y, features=x.shape[-1],
          name='expand')
      return nn.sigmoid(y) * x
```

**Review by marcvanzee — CHANGES_REQUESTED** [review]

**Comment by marcvanzee** [comment]
Sorry for the delay @joaogui1, I was OOO for 2 weeks. I left some comments, thanks for adding this!

File: docs/.vscode/settings.json [hunk 15]
```diff
@@ -0,0 +1,3 @@
+{
+    "restructuredtext.confPath": "${workspaceFolder}"
```
[the remark this answers is not in the document]
Comment (joaogui1) [reply docs/.vscode/settings.json:2]: it isn't, vscode keeps adding that everywhere :(

[the remark this answers is not in the document]
Comment (marcvanzee) [reply docs/.vscode/settings.json:2]: Could it be that you created a vscode workspace inside the "docs" folder? I think vscode will create a settings.json file for a workspace, and we added it to our `.gitignore. file in the top-level directory of our repository. So I would suggest you create a workspace from that directory, then this file won't be part of your commits.

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Looks great! If you remove the `settings.json` file it is ready to merge :-).

File: docs/.vscode/settings.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (joaogui1) [reply docs/.vscode/settings.json]: yeah, will do that from now on, thanks!

**Comment by joaogui1** [comment]
How can I locally run this test?

**Review by marcvanzee — APPROVED** [review]

**Comment by marcvanzee** [comment]
> How can I locally run this test?

From the top-level directory of our repo, run `sphinx-build -M doctest docs docs/_build`, which runs all tests in our docs.

They are also executed as part of our build test: https://github.com/google/flax/pull/1018/checks?check_run_id=2012580681 (section "Test testcode blocks in documentation")

**Comment by marcvanzee** [comment]
@joaogui1 I'm sorry about this, but I had to revert your HOWTO because of two reasons:

1. The code does not seem to use our Linen API. It does import `flax.linen`, but the code does not seem correct. For instance, this line:

```
nn.Dense(
          y, features=self.hidden_size,
          name='reduce')
```

Should be different in Linen:

```
nn.Dense(features=self.hidden_size, name='reduce')(x)
```

2. For the purpose of a HOWTO, I think it would be very insightful to our users to see how this class can actually be used. Did you use it yourself in any of your projects? The code doesn't have to be complex, but a simple demonstration seems useful since right now the code is not really showing "howto" to something, it is merely implementing a class. Another advantage of this is that it would immediately allow us to test the code better. Because the class is not used right now, it is also not tested.

Again, apologies for coming with this after you finished your PR, but we are in the process of defining what our HOWTOs should look like, which means we sometimes have to adjust our content a bit.

**Comment by joaogui1** [comment]
This was my mistake, I'll try to fix it today, do I have to open another PR or just push to this branch?

**Comment by marcvanzee** [comment]
> This was my mistake, I'll try to fix it today, do I have to open another PR or just push to this branch?

Thanks for fixing it! I think you can keep working on the same branch, but you will have to create a new PR since this one is merged already.
